File: Solutions by Programming Task/Fibonacci sequence/Python/Dictionary.py
D = {0: 0, 1: 1, 2: 1}
def Fib(x):
    if x in D:
        return D[x]
    f1 = Fib(x // 2 + 1)
    f2 = Fib((x - 1) // 2)
    D[x] = (f1 * f1 + f2 * f2 if x & 1 else f1 * f1 - f2 * f2)
    return D[x]

File: Solutions by Programming Task/Fibonacci sequence/Rust/Dictionary.rs
use std::collections::HashMap;

fn fib_impl(dict: &mut HashMap<u64, u64>, x: u64) -> u64 {
    if let Some(v) = dict.get(&x) {
        return *v
    }

    let v = match x {
        n if n < 2 => n,
        _ => fib_impl(dict, x - 1) + fib_impl(dict, x - 2)
    };

    dict.insert(x, v);
    v
}

fn fib(x: u64) -> u64 {
    let mut dict: HashMap<u64, u64> = (0..1).map(|x:u64| {(x, x)}).collect();
    fib_impl(&mut dict, x)
}

fn main() {
    println!("{}", fib(10));
}
